codecs resiliency, another approach

tmszdmsk opened the pull request:
After some thinking about #359 I prefer it this way because:
* it is more composable than writing the same logic in every codec
* covers all built-in and 3rd party codecs

Questions:
* can we assume that this should be the behavior for all possible `Injectors` and `Extractors`? Can you imagine any exception to this rule?
* it would be great to hide it from the external world, i.e. make it at least package scoped. IMHO they "belong" to `propagation` package naturally but if we want to make it "invisible" we can:
  * place them the main package where the `Tracer` with package scope
  * place it inside `Tracer`, but it is quite large already.
  * extract `PropagationRegistry` to a separate file and move decorators inside of it keeping it in the main package with package visibility. *I prefer this one*

I am new to this project so please think if I've missed something important from the design perspective.
 
To do
------
- [ ] tests

jpkrohling commented on jaeger-core/src/main/java/com/uber/jaeger/propagation/ExceptionCatchingExtractorDecorator.java:
```diff
@@ -0,0 +1,22 @@
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+@RequiredArgsConstructor
```
Could you add a javadoc, stating why this exists and how it works?

tmszdmsk replied:
you mean https://projectlombok.org/features/constructor?

jpkrohling replied:
No, sorry, I meant a class-level javadoc for the classes `ExceptionCatching*Decorator`

tmszdmsk replied:
sure, I will do that although I think that  `Injector` and `Extractor` interfaces need more attention in this area.

jpkrohling commented:
The build failed with `License violations were found: /.../jaeger-core/src/main/java/com/uber/jaeger/propagation/ExceptionCatchingExtractorDecorator.java,/home/travis/build/jaegertracing/jaeger-client-java/jaeger-core/src/main/java/com/uber/jaeger/propagation/ExceptionCatchingInjectorDecorator.java`

If you want to catch such errors faster, you may want to run `./gradlew check` locally before pushing updates to the PR :)

tmszdmsk commented:
@jpkrohling please, wait with low-level CR. I want to establish if it's an accepted way to go on a "design level" first.

tmszdmsk commented:
ok, I've extracted PropagationRegistry to separate class to hide exception-catching decorators from users. 

also, I've added javadoc proposition for `Extractor` which is based on my understanding of this topic. Please comment if I made wrong assumptions.

yurishkuro commented on jaeger-core/src/main/java/com/uber/jaeger/PropagationRegistry.java:
```diff
@@ -0,0 +1,65 @@
+
+  @SuppressWarnings("unchecked")
+  <T> Injector<T> getInjector(Format<T> format) {
+    return new ExceptionCatchingInjectorDecorator<T>((Injector<T>) injectors.get(format));
```
there's no reason to wrap the injector on every `get`, instead wrap it on `register`

yurishkuro commented on jaeger-core/src/main/java/com/uber/jaeger/PropagationRegistry.java:
```diff
@@ -0,0 +1,65 @@
+
+  @RequiredArgsConstructor
+  @Slf4j
+  public static class ExceptionCatchingExtractorDecorator<T> implements Extractor<T> {
```
why public?

tmszdmsk replied:
yep, automatic refactor ;)

yurishkuro commented on jaeger-core/src/main/java/com/uber/jaeger/PropagationRegistry.java:
```diff
@@ -0,0 +1,65 @@
+
+  @RequiredArgsConstructor
+  @Slf4j
+  public static class ExceptionCatchingInjectorDecorator<T> implements Injector<T> {
```
why public?

yurishkuro commented:
Thanks @tmszdmsk !

tmszdmsk commented:
no problem! ;)
